### ISSUE TITLE
Decorate for any step prefix instead of a single one (Given / When / Then)

### ISSUE THREAD

**Issue by Moult** [pytest-dev/pytest-bdd#450]
It would be nice to decorate, such as with `@step` which will decorate simultaneously for `Given`, and `When`, and `Then`.

**Comment by pohlt** [comment]
I strongly support this idea.

**Comment by elchupanebrej** [comment]
Please pay attention, that "pickles" produced by standard gherkin parser don't carry information about step type when steps is executed, so in case if the project would move to compatibility with Gherkin  tools it has to make steps keyword agnostic

**Comment by youtux** [comment]
Hi, what would be the use case for this change? If you bring a concrete example we can discuss it.

**Comment by pohlt** [comment]
Sure. An action could be a When step in some initial tests. Later on, the same action could be more of a preparation step, i.e., a Given:

```
Given the application is running
When some data is loaded
Then the patient information is displayed
```
Later on:
```
Given the application is running
And some data is loaded
When the user closes the application
Then the users is asked if the application should really be closed
```
For me, the Given step is just an indicator to the reader of a test which step is actually under investigation and the following Then step tests some assumptions. But in the end, almost any action can make sense as a Given, When, or even Then step.

**Comment by Moult** [comment]
I'm also using it when testing a longer more complex scenario broken up into two scenarios sometimes run separately:

```
Given the application is running
When some data is loaded
Then the patient information is displayed
```

```
Given the application is running
And some data is loaded
And the patient information is displayed
When the user does something
Then something else happens
```

**Comment by pohlt** [comment]
The current strict separation of step categories allows to associate different functions with the same "action string". I find it hard to come up with a good use case for this functionality, but I'm sure there must be some good reason.

**Comment by raspe88** [comment]
It would help keeping tests more maintainable. Taken the example from @pohlt: Without this feature a programmer would have to

1. introduce a new step description so it can be easily distinguished from other one's while they both mean the same
2. reimplement logic that has already been implemented for that new step

**Comment by elchupanebrej** [comment]
I've implemented this in my fork. Seems this will never be in pytest_bdd

**Comment by elchupanebrej** [comment]
This related to https://github.com/cucumber/common/pull/1741

**Comment by youtux** [comment]
Once https://github.com/pytest-dev/pytest-bdd/pull/540 is merged, you will be able to define your `step` function like this:

```python
def step(*step_args, **step_kwargs):
    caller_locals = sys._getframe(1).f_locals
    def decorator(fn):
        @given(*step_args, **step_kwargs, caller_locals=caller_locals)
        @when(*step_args, **step_kwargs, caller_locals=caller_locals)
        @then(*step_args, **step_kwargs, caller_locals=caller_locals)
        def wrapper(*args, **kwargs):
            return fn(*args, **kwargs)
        return wrapper
    return decorator
```

I don't want to provide this functionality out of the box in pytest-bdd, because I don't find it a good idea anyway, but at least you will be able to implement it and specialize it to your needs.

Fixed in #540.

**Comment by elchupanebrej** [comment]
@youtux https://github.com/cucumber/common/pull/1741 specifies the keyword of "Unspecified" type which could be handled by any of the step definitions. So there should be a mechanism in pytest-bdd to handle such steps.

Your answer for a `step` decorator also should be in the lib because from the start Gherkin hadn't specified any special context for every of its keyword. It was added by pytest-bdd itself, and because of this some of the integrations between Gherkin and pytest-bdd were not smooth.

**Comment by elchupanebrej** [comment]
Please check https://github.com/elchupanebrej/pytest-bdd-ng project, broad `step` function could be used to define steps for any keyword

**Comment by youtux** [comment]
Fixed by #548
